bittide/tests/Tests/DoubleBufferedRam.hs

```diff
@@ -885,3 +891,169 @@ wbStorageProtocolsModel = property $ do
+        WishboneMasterRequest addressWidth a ->
+        Bool
+    checkAddressEqual req1 req2 = extractAddress req1 == extractAddress req2
+        where
+            extractAddress :: WishboneMasterRequest addressWidth a -> BitVector addressWidth
+            extractAddress (Read addr _) = addr
+            extractAddress (Write addr _ _) = addr
+
+    getAddressWbRequest :: (KnownNat addressWidth, KnownNat (BitSize a)) =>
+        WishboneMasterRequest addressWidth a -> Unsigned addressWidth
+    getAddressWbRequest (Read addr _) = unpack addr
+    getAddressWbRequest (Write addr _ _) = unpack addr
+
+    getByteEnableWbRequest :: (KnownNat addressWidth, KnownNat (BitSize a)) =>
+        WishboneMasterRequest addressWidth a -> Unsigned (BitSize a `DivRU` 8)
+    getByteEnableWbRequest (Read _ be) = unpack be
+    getByteEnableWbRequest (Write _ be _) = unpack be
+
+-- | Memory Model
+type MemoryModel a b = Map.Map a b
+
+-- | Behavioral model for 'wbStorageTDPM'.
+wbStorageTDPBehaviorModel ::
+  forall addrW bytes .
+  ( 1 <= addrW, KnownNat bytes) =>
+  (KnownNat addrW) =>
+  [WishboneMasterRequest addrW (Bytes bytes)] -> [WishboneMasterRequest addrW (Bytes bytes)] ->
+  [(Transaction addrW bytes (Bytes bytes),Transaction addrW bytes (Bytes bytes))]
+wbStorageTDPBehaviorModel initWbOpsA initWbOpsB = case (cancelMulDiv @bytes @8) of
+  Dict -> snd $ L.mapAccumL g emptyMemoryMap (L.zipWith (\wbA wbB -> (wbA, wbB)) initWbOpsA initWbOpsB)
+   where
+    emptyMemoryMap = Map.empty :: MemoryModel (BitVector addrW) (Bytes bytes)
+
+    fromMaybeBytes :: Maybe (Bytes bytes) -> Bytes bytes
+    fromMaybeBytes x = case x of
+                    Just a  -> a
+                    Nothing -> 0
+
+    g storedMap (opA, opB) = (storedMapB, (transA, transB))
+      where
+        (storedMapA, transA) = f storedMap opA
+        (storedMapB, transB) = f storedMapA opB
+
+    -- Successful Read
+    f storedMap op@(Read i _) = (storedMap, ReadSuccess wbM2S wbS2M)
+     where
+      dat = fromMaybeBytes $ Map.lookup i storedMap
+      wbM2S = wbMasterRequestToM2S op
+      wbS2M = (emptyWishboneS2M @(Bytes bytes)){acknowledge = True, readData = dat}
+
+    -- Successful Write
+    f storedMap op@(Write i bs a) = (updatedStoredMap, WriteSuccess wbM2S wbS2M)
+     where
+      wbM2S = wbMasterRequestToM2S op
+      wbS2M = emptyWishboneS2M{acknowledge = True}
+      dat = fromMaybeBytes $ Map.lookup i storedMap
+      updatedStoredMap = Map.insert i (pack newEntry) storedMap
+
+      newEntry :: Vec bytes Byte
+      newEntry = zipWith3 (\ b old new -> if b then new else old)
+        (unpack bs) (unpack dat) (unpack a)
+
+-- | Check equiality of two BitVectors containing the undefined values ('.')
+-- Example:
+--  x: "000111..."
+--  y: "01.01.01."
+--  result: False
+eqBitVec :: (KnownNat n) => BitVector n -> BitVector n -> Bool
+eqBitVec x@(BV.BV xMask _) y =
+    case (x `xor` y) of
+        BV.BV m v -> xMask == m && v == 0
```

Comment:
Would https://hackage.haskell.org/package/clash-prelude-1.8.1/docs/Clash-Class-BitPack.html#v:isLike help you here?